lib/Dialect/Arc/Transforms/SplitFuncs.cpp

```diff
@@ -0,0 +1,164 @@
+};
+} // namespace
+
+void SplitFuncsPass::runOnOperation() {
+  symbolTable = &getAnalysis<SymbolTable>();
+  for (auto op : llvm::make_early_inc_range(getOperation().getOps<FuncOp>()))
+    if (failed(lowerFunc(op)))
+      return signalPassFailure();
+}
+
+LogicalResult SplitFuncsPass::lowerFunc(FuncOp funcOp) {
+  if (splitBound == 0)
+    return funcOp.emitError("Cannot split functions into functions of size 0.");
+  if (funcOp.getBody().getBlocks().size() > 1)
+    return funcOp.emitError("Regions with multiple blocks are not supported.");
+  assert(splitBound != 0 && "Cannot split functions into functions of size 0");
+  assert(funcOp->getNumRegions() == 1);
+  unsigned numOps = funcOp.front().getOperations().size();
+  if (numOps < splitBound)
+    return success();
+  int numBlocks = llvm::divideCeil(numOps, splitBound);
+  OpBuilder opBuilder(funcOp->getContext());
+  SmallVector<Block *> blocks;
+  Block *frontBlock = &(funcOp.getBody().front());
+  blocks.push_back(frontBlock);
+  for (int i = 0; i < numBlocks - 1; i++) {
+    SmallVector<Location> locs(frontBlock->getNumArguments(), funcOp.getLoc());
+    auto *block = opBuilder.createBlock(&(funcOp.getBody()), {},
+                                        frontBlock->getArgumentTypes(), locs);
+    blocks.push_back(block);
+  }
+
+  unsigned numOpsInBlock = 0;
+  SmallVector<Block *>::iterator blockIter = blocks.begin();
+  for (auto &op : llvm::make_early_inc_range(*frontBlock)) {
+    if (numOpsInBlock >= splitBound) {
+      blockIter++;
+      numOpsInBlock = 0;
+      opBuilder.setInsertionPointToEnd(*blockIter);
+    }
+    numOpsInBlock++;
+    // Don't bother moving ops to the original block
+    if (*blockIter == (frontBlock))
+      continue;
+    // Remove op from original block and insert in new block
+    op.remove();
+    (*blockIter)->push_back(&op);
+  }
+  DenseMap<Value, Value> argMap;
+  // Move function arguments to the block that will stay in the function
+  for (unsigned argIndex = 0; argIndex < frontBlock->getNumArguments();
+       argIndex++) {
+    auto oldArg = frontBlock->getArgument(argIndex);
+    auto newArg = blocks.back()->getArgument(argIndex);
+    replaceAllUsesInRegionWith(oldArg, newArg, funcOp.getBody());
+  }
+  Liveness liveness(funcOp);
+  auto argTypes = blocks.back()->getArgumentTypes();
+  auto args = blocks.back()->getArguments();
+
+  // Create return ops
+  for (int i = blocks.size() - 2; i >= 0; --i) {
+    liveness = Liveness(funcOp);
+    Block *currentBlock = blocks[i];
+    Liveness::ValueSetT liveOut = liveness.getLiveIn(blocks[i + 1]);
+    SmallVector<Value> outValues;
+    llvm::for_each(liveOut, [&outValues](auto el) {
+      if (!isa<BlockArgument>(el))
+        outValues.push_back(el);
+    });
+    opBuilder.setInsertionPointToEnd(currentBlock);
+    opBuilder.create<ReturnOp>(funcOp->getLoc(), outValues);
+  }
+  // Create and populate new FuncOps
+  for (long unsigned i = 0; i < blocks.size() - 1; i++) {
```

Comment:
Nit: inconsistent post and pre-increments, LLVM generally prefers pre-increment.